addons/delivery/tests/test_delivery_cost.py

```diff
@@ -353,6 +353,70 @@ def test_delivery_cost_gift_card(self):
+        # Set up the carrier
+        product_delivery = self.env['product.product'].create({
+            'name': 'Delivery Charges',
+            'type': 'service',
+            'list_price': 40.0,
+            'categ_id': self.env.ref('delivery.product_category_deliveries').id,
+        })
+        delivery_carrier = self.env['delivery.carrier'].create({
+            'name': 'Delivery Now Free Over 100',
+            'fixed_price': 40,
+            'delivery_type': 'fixed',
+            'invoice_policy': 'real',
+            'product_id': product_delivery.id,
+            'free_over': False,
+        })
+
+        so = self.SaleOrder.create({
+            'partner_id': self.partner_18.id,
+            'partner_invoice_id': self.partner_18.id,
+            'partner_shipping_id': self.partner_18.id,
+            'pricelist_id': self.pricelist.id,
+            'order_line': [(0, 0, {
+                'name': 'PC Assamble + 2GB RAM',
+                'product_id': self.product_4.id,
+                'product_uom_qty': 2,
+                'product_uom': self.product_uom_unit.id,
+                'price_unit': 120.00,
+            })],
+        })
+
+        delivery_wizard = Form(self.env['choose.delivery.carrier'].with_context({
+            'default_order_id': so.id,
+            'default_carrier_id': delivery_carrier.id
+        }))
+        delivery_wizard.save().button_confirm()
+
+        delivery_line = so.order_line.filtered(lambda line: line.is_delivery)
+        self.assertEqual(len(delivery_line), 1, "The delivery line is not present on the SO.")
+        self.assertEqual(delivery_line.price_unit, 0, "The delivery cost should be 0.")
+        so.action_confirm()
+
+        picking = so.picking_ids[0]
+        self.assertEqual(picking.carrier_id.id, so.carrier_id.id, "Carrier is not the same on Picking and on SO.")
+        picking.move_ids[0].quantity_done = 1.0
+        self.assertGreater(picking.shipping_weight, 0.0, "Picking weight should be positive.")
+
+        # Confirm picking for one quantiy and create a back order for the second
+        picking._action_done()
+        self.assertEqual(picking.carrier_price, 40.0, "Price is not set")
+        # Check that the delivery cost (previously set to 0) has been correctly updated
+        self.assertEqual(picking.carrier_price, delivery_line.price_unit, "The delivery cost is not updated")
+
+        # confirm the back order
+        bo = picking.backorder_ids
+        bo.move_ids[0].quantity_done = 1.0
+        self.assertGreater(bo.shipping_weight, 0.0, "Picking weight should be positive.")
+        bo._action_done()
+        self.assertEqual(bo.carrier_price, 40.0, "Price is not set")
+
+        new_delivery_line = so.order_line.filtered(lambda line: line.is_delivery) - delivery_line
+        self.assertEqual(len(new_delivery_line), 1, "One new line should be created")
+        self.assertEqual(bo.carrier_price, new_delivery_line.price_unit, "The delivery cost is not updated")
```

Comment:
Except if I'm wrong, you are fixing the PU of the SOL (you ensure that the `carrier_price` is correctly propagated)
So, you should reverse the assert arguments: first is what we test, second is what we want
```suggestion
        self.assertEqual(new_delivery_line.price_unit, bo.carrier_price)
```